packages/loopring_v3/contracts/aux/bridge/Bridge.sol

```diff
@@ -0,0 +1,902 @@
+            uint tokenID      = (packedData >> 88) & 0xffff;
+            uint amount       = (packedData >> 64) & 0xffffff;
+            uint fee          = (packedData >> 32) & 0xffff;
+            // Decode floats
+            amount = (amount & 524287) * (10 ** (amount >> 19));
+            fee = (fee & 2047) * (10 ** (fee >> 11));
+
+            // Verify the transaction data
+            require(
+                // txType == ExchangeData.TransactionType.TRANSFER &&
+                // transfer.type == 1 &&
+                // transfer.fromAccountID == ctx.accountID &&
+                // transfer.toAccountID == UNKNOWN  &&
+                packedData & 0xffffffffffff0000000000000000000000000000000000 ==
+                (uint(ExchangeData.TransactionType.TRANSFER) << 176) | (1 << 168) | (uint(accountID) << 136) &&
+                /*feeTokenID*/(packedData >> 48) & 0xffff == tokenID &&
+                fee <= (amount * MAX_FEE_BIPS / 10000) &&
+                (100000 - 8) * targetAmount <= 100000 * amount && amount <= targetAmount,
+                "INVALID_BRIDGE_TRANSFER_TX_DATA"
+            );
+
+            // Pack the transfer data to compare against batch deposit hash
+            assembly {
+                mstore(add(transfers, 2), tokenID)
+                mstore(    transfers    , or(shl(96, to), targetAmount))
+                transfers := add(transfers, 34)
+            }
+        }
+
+        // Get the original transfers ptr back
+        assembly {
+            transfers := sub(transfers, add(32, mul(34, mload(amounts))))
+        }
+        // Check if these transfers can be processed
+        bytes32 hash = _hashTransferBatch(transfers);
+        require(!_arePendingTransfersTooOld(batch.batchID, hash), "TRANSFERS_TOO_OLD");
+
+        // Mark transfers as completed
+        delete pendingTransfers[batch.batchID][hash];
+    }
+
+    function _processBridgeCalls(
+        Context          memory   ctx,
+        ConnectorCalls[] calldata connectorCalls
+        )
+        internal
+    {
+        // Total amounts transferred to the bridge
+        uint[] memory totalAmounts = new uint[](ctx.tokens.length);
+
+        // All resulting deposits from all bridge calls
+        BridgeTransfer[][] memory deposits = new BridgeTransfer[][](connectorCalls.length);
+
+        // Verify and execute bridge calls
+        for (uint c = 0; c < connectorCalls.length; c++) {
+            ConnectorCalls calldata connectorCall = connectorCalls[c];
+
+            // Verify the transactions
+            _processConnectorCall(
+                ctx,
+                connectorCall,
+                totalAmounts
+            );
+
+            // Call the connector
+            deposits[c] = _connectorCall(
+                ctx,
+                connectorCall,
+                c,
+                connectorCalls
+            );
+        }
+
+        // Verify withdrawals
+        _processWithdrawals(ctx, totalAmounts);
```

Comment:
@Brechtpd shall we process withdrawals before calling connectors?

Reply:
Putting this here for other people looking at this PR: The bridge transactions are processed *after* the block has already been processed, which means all withdrawals on L1 already happened. So the funds are already in the Bridge contract before any of these transactions are even verified. So the order for L1 doesn't matter.
For the L2 transactions we do have to first process all transfers to the Bridge before we can withdraw from the Bridge, otherwise the Bridge accounts wouldn't contain any funds so the withdrawal would be invalid on L2.